src/game/client/maputil.cpp

```diff
@@ -67,11 +94,848 @@ Utf8String MapCache::Get_User_Map_Dir()
+            }
+        }
+
+        return selected;
+    } else {
+        return -1;
+    }
+}
+
+int Populate_Map_List_Box(GameWindow *listbox, bool is_official, bool is_multiplayer, Utf8String map_name)
+{
+    if (g_theMapCache != nullptr && listbox != nullptr) {
+        Gadget_List_Box_Reset(listbox);
+        return Populate_Map_List_Box_No_Reset(listbox, is_official, is_multiplayer, map_name);
+    } else {
+        return -1;
+    }
+}
+
+bool Is_Valid_Map(Utf8String map_name, bool is_multiplayer)
+{
+    if (g_theMapCache != nullptr && !map_name.Is_Empty()) {
+        g_theMapCache->Update_Cache();
+        map_name.To_Lower();
+        auto it = g_theMapCache->find(map_name);
+        if (it != g_theMapCache->end() && is_multiplayer == it->second.m_isMultiplayer) {
+            return true;
+        } else {
+            return false;
+        }
+    } else {
+        return false;
+    }
+}
+
+Utf8String Get_Default_Map(bool is_multiplayer)
+{
+    if (g_theMapCache != nullptr) {
+        g_theMapCache->Update_Cache();
+
+        for (auto it = g_theMapCache->begin(); it != g_theMapCache->end(); it++) {
+            if (is_multiplayer == it->second.m_isMultiplayer) {
+                return it->first;
+            }
+        }
+
+        return Utf8String::s_emptyString;
+    }
+
+    return Utf8String::s_emptyString;
+}
+
+Utf8String Get_Official_Multiplayer_Map()
+{
+    if (g_theMapCache != nullptr) {
+        g_theMapCache->Update_Cache();
+
+        for (auto it = g_theMapCache->begin(); it != g_theMapCache->end(); it++) {
+            if (it->second.m_isMultiplayer && it->second.m_isOfficial) {
+                return it->first;
+            }
+        }
+
+        return Utf8String::s_emptyString;
+    }
+
+    return Utf8String::s_emptyString;
+}
+
+bool Is_Map_Official(Utf8String map_name)
+{
+    if (g_theMapCache != nullptr && !map_name.Is_Empty()) {
+        g_theMapCache->Update_Cache();
+        map_name.To_Lower();
+        auto it = g_theMapCache->find(map_name);
+        if (it != g_theMapCache->end()) {
+            return it->second.m_isOfficial;
+        } else {
+            return false;
+        }
+    } else {
+        return false;
+    }
+}
+
+void Copy_From_Big_To_Dir(Utf8String &source, Utf8String &dest)
+{
+    File *file = g_theFileSystem->Open_File(source.Str(), File::READ | File::BINARY);
+    captainslog_relassert(file != nullptr, 6, "Copy_From_Big_To_Dir - Error opening source file '%s'", source.Str());
+    int size = file->Seek(0, File::END);
+    file->Seek(0, File::START);
+    char *buf = new char[size];
+    captainslog_relassert(buf != nullptr, 6, "Copy_From_Big_To_Dir - Unable to allocate buffer for file '%s'", source.Str());
+    int read_size = file->Read(buf, size);
+    captainslog_relassert(read_size >= size, 6, "Copy_From_Big_To_Dir - Error reading from file '%s'", source.Str());
```

Comment:
Should be `read_size == size` no? If `read_size` was larger than `size`, then it would have written over buffer end.

Reply:
This is what the original code does.